### ISSUE TITLE
tune2fs man page should have more warnings about large_dir

### ISSUE THREAD

**Issue by jmtilli** [tytso/e2fsprogs#105]
I have managed to make my Linux system unbootable twice in the last 3 years.

The first time, I didn't have any idea why it happened. It stuck in grub. I managed to boot on a rescue disk and mount the filesystem and copy all important files somewhere else.

The second time, it happened last Friday. Grub wouldn't recognise my filesystem. This time, I tried to repair grub but it wouldn't work. Quick googling suggested that some ext2fs filesystem options don't work with grub. Then I remembered I wanted to create a directory with more than 65000 subdirectories and found large_dir, and then found that wasn't what I wanted and found dir_nlink.

I tried to remove large_dir from the filesystem but tune2fs wouldn't let me do that. So I had to copy all files somewhere else and reinstall.

Currently tune2fs man page says about large_dir:

                   large_dir
                          Increase the limit on the number of files per direc‐
                          tory.   Tune2fs currently only supports setting this
                          filesystem feature.

...which doesn't suggest using it might be a bad idea.

I think the man page should distinctly warn that (1) grub won't boot from any filesystem with large_dir, (2) you cannot disable large_dir once enabled so it's impossible to make it re-bootable again after enabling it (now it says that in a manner that doesn't suggest enabling large_dir would be very dangerous), and (3) in some cases the user might want to enable dir_nlink instead of large_dir.

It might even make sense to have a "do you really want to make your system unbootable?" question in tune2fs itself, in case someone finds large_dir on some website such as Stack Overflow and enables it without reading the man page.

**Comment by tytso** [comment]
I'm sorry you've managed to make your system not bootable.   Out of curiosity, why were you trying to enable large_dir in the first place?   Are you cutting and pasting from some receipe to turn on all ext4 features?   (While we do appreciate guinea pigs not all file system features are equally stable, and at the FAST conference poster session a graduate student noted that there are 10**37 different configuration states, and they were trying to create an automated system using LLVM and corepath analysis to figure out a set of configuration options that maximizes test coverage.)

The problem with adding a man page is that hopefully such a warning will be out of date.    I've already submitted a trivial patch to grub2 to allow it to support the large_dir feature: https://savannah.gnu.org/bugs/?61606.   Also, not all systems use the grub2 boot loader, and this is really a shrotcoming in grub2 (and it's partially because grub2 insists on using GPL3 and so has a license compatibility issue with libext2fs).

Note that unless you are creating ***really*** gargantuan numbers of directory entries in a directory, you really don't need the large_dir feature.   It was originally contributed by the folks who work on the lustre clsuter file system, where they were storing nearly all of their objects in a single directory, and the file system was a very large RAID array.  In such systems, it's super rare that the boot device is on the large data file system.   Using a single root file system is typical for home / personal systems, and these generally aren't creating directories which have enough entries such that the directory is larger than ~450MB (or more than 16 million files assuming an average file name size of 16 bytes).

**Comment by jmtilli** [comment]
Yes, certainly it's far better to support large_dir in grub rather than to add a warning into the manual page. So if there's already a patch for grub, I agree that patch should be merged instead.

The reason for enabling large_dir was that I somehow was led to believe that it's the feature that would allow more than 65000 subdirectories in a directory (although dir_nlink was really that feature). I didn't enable every single ext4 feature in a hope to make my system "work better".
